### PR TITLE
[v1.x] Static build for mxnet-cu110

### PR DESCRIPTION
**Pull request by waytrue17**
## Description ##
Added static build MXNet with CUDA 11.0

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)

### Changes ###
- [x] added script to statically build MXNet with CUDA 11.0

### CODE REVIEW

**Comment by waytrue17** [comment]
@leezu @ChaiBapchya Can you please review this PR?

**Review by josephevans — APPROVED** [review]

**Comment by sandeep-krishnamurthy** [comment]
Thank you so much @waytrue17

File: tools/pip/setup.py [hunk 1]
```diff
@@ -215,4 +217,4 @@ def skip_markdown_comments(md):
           'Topic :: Software Development :: Libraries',
           'Topic :: Software Development :: Libraries :: Python Modules',
       ],
-      url='https://github.com/apache/incubator-mxnet')
+      url='https://github.com/apache/incubator-mxnet')
```
Comment (ChaiBapchya) [inline tools/pip/setup.py:220]: nitpick: newline at the end of the file.

File: tools/setup_gpu_build_tools.sh [hunk 2]
```diff
@@ -312,11 +345,16 @@ if [[ ! -d $DEPS_PATH/usr/local/cuda-${CUDA_MAJOR_VERSION} ]]; then
...
+            newfile=$(basename $h | sed 's/_v8//')
+            cp $h ${prefix}/include/$newfile
+        done
+    fi
+    cp -f ${prefix}/usr/include/x86_64-linux-gnu/cudnn_v${LIBCUDNN_MAJOR}.h ${prefix}/include/cudnn.h
+    ln -sf libcudnn_static_v${LIBCUDNN_MAJOR}.a ${prefix}/usr/lib/x86_64-linux-gnu/libcudnn.a
+    cp -f ${prefix}/usr/local/cuda-${CUDA_MAJOR_VERSION}/lib64/*.a ${prefix}/lib/
+    cp -f ${prefix}/usr/include/nccl.h ${prefix}/include/nccl.h
+    ln -sf libnccl_static.a ${prefix}/usr/lib/x86_64-linux-gnu/libnccl.a
+fi
```
Comment (ChaiBapchya) [inline tools/setup_gpu_build_tools.sh:360]: nitpick: new line

**Comment by ChaiBapchya** [comment]
How & Where has this change been tested?

File: tools/setup_gpu_build_tools.sh (hunk 2 above)
Comment (waytrue17) [reply tools/setup_gpu_build_tools.sh]: New line added

File: tools/pip/setup.py (hunk 1 above)
Comment (waytrue17) [reply tools/pip/setup.py]: New line added

**Comment by waytrue17** [comment]
> How & Where has this change been tested?

I tested it by calling `tools/staticbuild/build.sh cu110` to build `libmxnet.so` on a gpu instance, binding it with python language. Then run a simple script in python with `ctx=mx.gpu(0)`.  
The above tested was performed on v1.8.x with the same change #19273.

**Review by ChaiBapchya — APPROVED** [review]
Sure. thanks & hope we confirmed `ldd libmxnet.so` is linked to cuda11 so

**Comment by waytrue17** [comment]
> Sure. thanks & hope we confirmed `ldd libmxnet.so` is linked to cuda11 so

@ChaiBapchya Thanks for the suggestion. The following is the `ldd libmxnet.so` outputs:
`linux-vdso.so.1 (0x00007ffda2bb9000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f5f8c986000)
	libcudart.so.11.0 => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudart.so.11.0 (0x00007f5f8c708000)
	libcublas.so.11 => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libcublas.so.11 (0x00007f5f868b8000)
	libcurand.so.10 => /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcurand.so.10 (0x00007f5f82857000)
	libcusolver.so.10 => /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcusolver.so.10 (0x00007f5f77d4c000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f5f77b44000)
	libnvToolsExt.so.1 => /usr/local/cuda/lib64/libnvToolsExt.so.1 (0x00007f5f7793b000)
	libgfortran.so.4 (0x00007f5f7755c000)
	libcufft.so.10 => /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcufft.so.10 (0x00007f5f6ef22000)
	libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007f5f6da6f000)
	libnvrtc.so.11.0 => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libnvrtc.so.11.0 (0x00007f5f6c284000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f5f6befb000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5f6bb5d000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f5f6b92e000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5f6b716000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5f6b4f7000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5f6b106000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5fe6faa000)
	libcublasLt.so.11 => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libcublasLt.so.11 (0x00007f5f5ff77000)
	libquadmath.so.0 (0x00007f5f5fd37000)
`

**Comment by szha** [comment]
@waytrue17 could you also add this to master?

**Review by szha — COMMENTED** [review]
A couple more missing items:
- CMake config in https://github.com/apache/incubator-mxnet/tree/v1.x/config/distribution
- pip package description doc in https://github.com/apache/incubator-mxnet/tree/v1.x/tools/pip/doc

**Comment by szha** [comment]
Remember to also configure the CD pipeline for a nightly build of this variant for testing.

File: tools/setup_gpu_build_tools.sh [hunk 3]
```diff
@@ -23,11 +23,22 @@
...
 
 >&2 echo "Setting CUDA versions for $VARIANT"
-if [[ $VARIANT == cu102* ]]; then
+if [[ $VARIANT == cu110* ]]; then
+    CUDA_VERSION='11.0.221-1'
+    CUDA_PATCH_VERSION='11.2.0.252-1'
+    CUDA_LIBS_VERSION='10.2.1.245-1'
+    CUDA_SOLVER_VERSION='10.6.0.245-1'
+    CUDA_NVTX_VERSION='11.0.167-1'
+    LIBCUDA_VERSION='450.51.06-0ubuntu1'
```
Comment (szha) [inline tools/setup_gpu_build_tools.sh:38]: is this the minimum required version? @DickJC123

**Comment by szha** [comment]
Looks like CUDA 11.1 is available now and we should consider adding that too.

**Comment by waytrue17** [comment]
> @waytrue17 could you also port this to master once merged? It will help resolve [#18657 (comment)](https://github.com/apache/incubator-mxnet/issues/18657#issuecomment-697985192)

opened a separate PR for master branch:https://github.com/apache/incubator-mxnet/pull/19295

**Comment by waytrue17** [comment]
> A couple more missing items:
> 
> * CMake config in https://github.com/apache/incubator-mxnet/tree/v1.x/config/distribution
> * pip package description doc in https://github.com/apache/incubator-mxnet/tree/v1.x/tools/pip/doc

Added the missing items and also configured the cd tests. Please review again, thanks

**Comment by waytrue17** [comment]
@mxnet-bot run ci [centos-gpu]

File: tools/setup_gpu_build_tools.sh (hunk 3 above)
Comment (szha) [reply tools/setup_gpu_build_tools.sh:38]: I think you can use 450.36.06 here instead. @waytrue17

Comment (waytrue17) [reply tools/setup_gpu_build_tools.sh]: @szha Updated the version, please review again, thanks!

**Review by szha — APPROVED** [review]

File: make/staticbuild/linux_cu110.mk [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+USE_CUDA_PATH = $(DEPS_PATH)/usr/local/cuda-11.0
+
+# whether to use CuDNN library
+USE_CUDNN = 1
+
+# whether to use NCCL library
+USE_NCCL = 1
+
+# CUDA architecture setting: going with all of them.
+# For CUDA < 6.0, comment the *_50 lines for compatibility.
```
Comment (mseth10) [inline make/staticbuild/linux_cu110.mk:78]: hey @waytrue17 have we tried building using this make config? how do we ensure we do not compile arch=compute_30 here? We had to remove "3.0" from MXNET_CUDA_ARCH in the cmake config file. Is there something similar we need to do here.

Comment (waytrue17) [reply make/staticbuild/linux_cu110.mk:78]: Hi @mseth10, I have tested it on EC2 gpu instance and we also used it in our aws-mxnet-cu110 codebuild pipeline.

**Comment by waytrue17** [comment]
@mxnet-bot run ci [sanity]

**Review by mseth10 — APPROVED** [review]
LGTM

**Pull request merged**